src/monitor/mod.rs

```diff
@@ -0,0 +1,570 @@
+            Self::NodeOutsideCurrentScope(node, scope) => write!(
+                f,
+                "node id {} does not belong to the current scope {}",
+                node, scope
+            ),
+            Self::UnexpectedNodeId => {
+                f.write_str("node id is not valid for this event type in the current state")
+            }
+            Self::NoRegion => f.write_str("PopRegion event without matching PushRegion"),
+            Self::InvalidEvent(descr) => f.write_str(descr),
+        }
+    }
+}
+
+/// `TraceMonitor` listens to and validates event traces emitted by a
+/// circuit.
+///
+/// The monitor uses [`CircuitEvent`](`crate::circuit::trace::CircuitEvent`)s,
+/// to build up a model of the circuit graph and check that nodes aren't added
+/// twice, the parent circuit exists when a node is added, edges connect
+/// existing nodes, etc.
+///
+/// During circuit evaluation, the monitor uses
+/// [`SchedulerEvent`](`crate::circuit::trace::SchedulerEvent`)s, to track the
+/// state of the circuit automaton (see
+/// [`SchedulerEvent`](`crate::circuit::trace::SchedulerEvent`) and detect
+/// invalid event sequences.  It uses the circuit graph constructed from
+/// `CircuitEvent`s to check that events relate to existing nodes and match the
+/// type of each node (e.g., the
+/// [`ClockStart`](`crate::circuit::trace::SchedulerEvent::ClockStart`) event
+/// can only be emitted by an iterative circuit).
+///
+/// The monitor is initialized with a pair of callbacks to invoke when an
+/// invalid `CircuitEvent` or `SchedulerEvent` is observed.  These
+/// callbacks can simply panic when using the trace monitor in `cargo
+/// test`.
+#[repr(transparent)]
+#[derive(Clone)]
+pub struct TraceMonitor(Arc<Mutex<TraceMonitorInternal>>);
+
+impl TraceMonitor {
+    /// Attach trace monitor to a circuit.  The monitor will register for
+    /// both `CircuitEvent`s and `SchedulerEvent`s.
+    pub fn attach(&self, circuit: &Circuit<()>, handler_name: &str) {
+        TraceMonitorInternal::attach(self.0.clone(), circuit, handler_name);
+    }
+
+    pub fn new<CE, SE>(circuit_error_handler: CE, scheduler_error_handler: SE) -> Self
+    where
+        CE: Fn(&CircuitEvent, &TraceError) + 'static,
+        SE: Fn(&SchedulerEvent, &TraceError) + 'static,
+    {
+        Self(Arc::new(Mutex::new(TraceMonitorInternal::new(
+            circuit_error_handler,
+            scheduler_error_handler,
+        ))))
+    }
+
+    pub fn new_panic_on_error() -> Self {
+        Self(Arc::new(Mutex::new(
+            TraceMonitorInternal::new_panic_on_error(),
+        )))
+    }
+
+    pub fn visualize_circuit(&self) -> VisGraph {
+        self.visualize_circuit_annotate(&|_| "".to_string())
+    }
+
+    pub fn visualize_circuit_annotate<F>(&self, f: &F) -> VisGraph
+    where
+        F: Fn(&GlobalNodeId) -> String,
+    {
+        self.0.lock().unwrap().circuit.visualize(f)
+    }
```

Comment:
This shouldn't take a reference to a closure and should take a `FnMut` to give greater flexibility. I'd also like to see a better name than `f` for the argument and some docs on what the closure is for
```suggestion
    pub fn visualize_circuit_annotate<F>(&self, f: F) -> VisGraph
    where
        F: FnMut(&GlobalNodeId) -> String,
    {
        self.0.lock().unwrap().circuit.visualize(f)
    }
```